server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -325,6 +347,137 @@ public IndexInput openInput(String name, IOContext context) throws IOException {
+
+        List<CompletableFuture<Void>> resultFutures = new ArrayList<>();
+
+        boolean uploadOfAllFilesSuccessful = true;
+        for (String src : files) {
+            String remoteFilename = createRemoteFileName(src, false);
+            uploadTracker.beforeUpload(src);
+            if (remoteDataDirectory.getBlobContainer().isMultiStreamUploadSupported()) {
+                try {
+                    CompletableFuture<Void> resultFuture = createUploadFuture(from, src, remoteFilename, context);
+                    resultFuture.whenComplete((uploadResponse, throwable) -> {
+                        if (throwable != null) {
+                            uploadTracker.onFailure(src);
+                        } else {
+                            uploadTracker.onSuccess(src);
+                        }
+                    });
+                    resultFutures.add(resultFuture);
+                } catch (Exception e) {
+                    uploadTracker.onFailure(src);
+                    throw e;
+                }
+            } else {
+                boolean success = true;
+                try {
+                    copyFrom(from, src, src, context, false);
+                    uploadTracker.onSuccess(src);
+                } catch (IOException e) {
+                    success = false;
+                    uploadOfAllFilesSuccessful = false;
+                    logger.warn(() -> new ParameterizedMessage("Exception while uploading file {} to the remote segment store", src), e);
+                } finally {
+                    if (!success) {
+                        uploadTracker.onFailure(src);
+                    }
+                }
+            }
+        }
+
+        if (resultFutures.isEmpty() == false) {
+            CompletableFuture<Void> resultFuture = CompletableFuture.allOf(resultFutures.toArray(new CompletableFuture[0]));
+            try {
+                resultFuture.get();
+            } catch (ExecutionException e) {
+                IOException corruptIndexException = ExceptionsHelper.unwrapCorruption(e);
+                if (corruptIndexException != null) {
+                    throw corruptIndexException;
+                }
+                Throwable throwable = ExceptionsHelper.unwrap(e, CorruptFileException.class);
+                if (throwable != null) {
+                    CorruptFileException corruptFileException = (CorruptFileException) throwable;
+                    throw new CorruptIndexException(corruptFileException.getMessage(), corruptFileException.getFileName());
+                }
+                throw e;
+            }
+        }
+
+        return uploadOfAllFilesSuccessful;
+    }
+
+    private CompletableFuture<Void> createUploadFuture(Directory from, String src, String remoteFileName, IOContext ioContext)
+        throws Exception {
+
+        AtomicReference<Exception> exceptionRef = new AtomicReference<>();
+        long expectedChecksum = calculateChecksumOfChecksum(from, src);
```

Comment:
How are we calculating checksum of the entire file here ?

Reply:
@gbbafna , lets think about a segment file as composed of two parts
1. Actual lucene content
2. Checksum of the stored lucene content
For s3 upload, we need the checksum of the entire file i.e. checksum of lucene content + checksum of the stored file.
We don't need to calculate the checksum of lucene content since we already have it in the footer. All we need to calculate is the checksum of the stored checksum in the footer of the file. We do it and then combine both checksums which gives us the checksum of the entire file.